src/main/java/com/aws/greengrass/mqttclient/AwsIotMqttClient.java

```diff
@@ -227,14 +245,67 @@ int getTimeout() {
                 MqttClient.DEFAULT_MQTT_OPERATION_TIMEOUT, MqttClient.MQTT_OPERATION_TIMEOUT_KEY));
     }
 
-    private void resubscribe() {
-        subscriptionTopics.forEach((key, value) -> {
-            subscribe(key, value).whenComplete((i, t) -> {
-                if (t != null) {
-                    logger.atError().kv(TOPIC_KEY, key).kv(QOS_KEY, value.name()).log("Unable to resubscribe to topic");
+    /**
+     * Run re-subscription task in another thread so that the current thread is not blocked by it. The task will keep
+     * retrying until all subscription succeeded or it's canceled by network interruption.
+     *
+     * @param sessionPresent whether the session persisted
+     */
+    private synchronized void resubscribe(boolean sessionPresent) {
+        // Don't bother spin up a thread if subscription is empty
+        if (!subscriptionTopics.isEmpty()) {
+            // If connected without a session, all subscriptions are dropped and need to be resubscribed
+            if (!sessionPresent) {
+                droppedSubscriptionTopics.putAll(subscriptionTopics);
+            }
+            if (!droppedSubscriptionTopics.isEmpty() && (resubscribeFuture == null || resubscribeFuture.isDone())) {
+                resubscribeFuture = executorService.submit(this::resubscribeDroppedTopicsTask);
+            }
+        }
+    }
+
+    private void resubscribeDroppedTopicsTask() {
+        long delayMillis = 0;  // don't delay the first run
+        while (!droppedSubscriptionTopics.isEmpty()) {
+            logger.atDebug().event(RESUB_LOG_EVENT).kv("droppedTopics", droppedSubscriptionTopics.keySet())
+                    .kv("delayMillis", delayMillis).log("Subscribing to dropped topics");
+            ScheduledFuture<?> scheduledFuture = ses.schedule(() -> {
```

Comment:
What's the reason of using `ses`? Would we end up scheduling multiple resubscribe tasks?

Reply:
We won't schedule multiple resub tasks because later we call `future.get` to wait for it. So this is actually equivalent to `while { ... Thread.sleep }`. I thought waiting with `Future.get` instead of `Thread.sleep` has better performance (I haven't done experiment though). Downside is that the code is much complicated.